apps/server/lib/lexical/server/code_intelligence/entity.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Lexical.Server.CodeIntelligence.Entity do
+  alias Future.Code, as: Code
+  alias Lexical.Ast
+  alias Lexical.Document
+  alias Lexical.Document.Location
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Lexical.Project
+  alias Lexical.RemoteControl
+  alias Lexical.Text
+
+  require Logger
+
+  @type resolved :: {:module, module()}
+
+  @doc """
+  Attempts to resolve the entity at the given position in the document.
+
+  ## Return values
+
+  Returns `{:ok, resolved, range}` if successful and `{:error, error}`
+  otherwise. The `range` includes the resolved node and the
+  Resolved entities are one of:
+
+    * `{:module, module}`
+
+  """
+  @spec resolve(Document.t(), Position.t()) :: {:ok, resolved, Range.t()} | {:error, term()}
+  def resolve(%Document{} = document, %Position{} = position) do
+    with {:ok, %{context: context, begin: begin_pos, end: end_pos}} <-
```

Comment:
can ast return anything other than this? I think matching out all the pieces is a little noisy. 
I'm also wondering if we should wrap the map in a struct. I really really dislike maps as data structures. You have no idea what's in them.

Maybe this is for a subsequent review.

Reply:
> can ast return anything other than this?

It can also return `{:ok, :none}` or `{:error, parse_error}`.

Right now it's a pretty thin wrapper over `Code.Fragment.surround_context/3`. We can play with the return value, however. Instead of a struct, I'd be more inclined to return `{:ok, context, range}` and turn the `:none` case into an `{:error, ...}` of some kind. But, doing that, we may want to change the name of the function as well so that there isn't an expectation that it'll return the same thing as `Code.Fragment.surround_context`.

Reply:
Will address this in followup PR.